src/Database/LSMTree/Internal.hs

```diff
@@ -1390,21 +1404,12 @@ duplicate th = do
           -- The table contents escape the read access, but we just added references
           -- to each run so it is safe.
           content <- RW.withReadAccess (tableContent thEnv) $ \content -> do
-            forRunM_ (tableLevels content) $ \r -> do
-              allocateTemp reg
-                (Run.addReference r)
-                (\_ -> Run.removeReference r)
+            addReferenceTableContent reg content
             pure content
-          WBB.addReference (tableWriteBufferBlobs content)
-          -- TODO: Fix possible double-free! See 'newCursor'.
-          -- In `newWith`, the table handle (in the open state) gets added to
-          -- `sessionOpenTables', even if later an async exception occurs and
-          -- the temp registry rolls back all allocations.
           newWith
+            reg
```

Comment:
I think it may be clearer to change lines 1397 and 1399 to `th{..}` and `\thEnv{..}`, respectively,  rather than explicitly calling the record accessor functions here.